### PR TITLE
Make frontend build in Docker image optional

### PR DESCRIPTION
**Pull request by arikfr**
## What type of PR is this? (check all applicable)

- [x] Other

## Description

A different approach to making frontend build optional by using a build arg. While a bit more hacky, it's fully supported across different Docker versions and requires less changes.

## Related Tickets & Documents

#4843, #4292

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,20 +1,25 @@
...
 WORKDIR /frontend
 COPY package.json package-lock.json /frontend/
 COPY viz-lib /frontend/viz-lib
-RUN npm ci --unsafe-perm
+RUN if [ "x$skip_frontend_build" = "x" ] ; then npm ci --unsafe-perm; fi
 
 COPY client /frontend/client
 COPY webpack.config.js /frontend/
-RUN npm run build
+RUN if [ "x$skip_frontend_build" = "x" ] ; then npm run build; else mkdir /frontend/client/dist && touch /frontend/client/dist/{multi_org.html,index.html}; fi
```
Comment (arikfr) [inline Dockerfile:13]: Create dummy `index.html` and `multi_org.html` to make backend tests work.

File: Dockerfile [hunk 2]
```diff
@@ -1,20 +1,25 @@
...
 
 FROM python:3.7-slim
 
 EXPOSE 5000
 
 # Controls whether to install extra dependencies needed for all data sources.
 ARG skip_ds_deps
+# Controls whether to build the frontend assets
+ARG skip_frontend_build
```
Comment (arikfr) [inline Dockerfile:22]: This isn't used and will be removed in a future commit.

File: docker-compose.yml [hunk 3]
```diff
@@ -1,12 +1,18 @@
...
+    context: .
+    args:
+      skip_frontend_build: "true"
   volumes:
     - .:/app
 x-redash-environment: &redash-environment
+  REDASH_LOG_LEVEL: "INFO"
   REDASH_REDIS_URL: "redis://redis:6379/0"
+  REDASH_DATABASE_URL: "postgresql://postgres@postgres/postgres"
+  REDASH_RATELIMIT_ENABLED: "false"
```
Comment (arikfr) [inline docker-compose.yml:15]: This isn't directly related, but used the opportunity to reduce even more duplicates.

**Comment by koooge** [comment]
I think it's a reasonable proposal. On another note, my works care about no-include requirements_dev.txt for production build. Is it possible in this PR?

Anyway, Makefile is no longer update? #3038

**Comment by arikfr** [comment]
> I think it's a reasonable proposal. On another note, my works care about no-include requirements_dev.txt for production build. Is it possible in this PR?

https://github.com/getredash/redash/pull/4879/commits/202e1da0f126feb5bb2b676f7bdf72fb676a80fd

> By the way, Makefile is no longer update? #3038

I never used the Makefile... It's possible it's a bit out of sync.

File: Dockerfile [hunk 4]
```diff
@@ -70,6 +72,7 @@ ENV PIP_NO_CACHE_DIR=1
 # change.
 COPY requirements.txt requirements_bundles.txt requirements_dev.txt requirements_all_ds.txt ./
 RUN pip install -r requirements.txt -r requirements_dev.txt
```
Comment (koooge) [inline Dockerfile:74]: Please delete this line.

Comment (arikfr) [reply Dockerfile:74]: Oops. 🤦 

```suggestion
```

File: docker-compose.yml (hunk 3 above)
Comment (jezdez) [reply docker-compose.yml:15]: I think we had to fix the scheduler container for this anyway in our recent rebase: https://github.com/mozilla/redash/commit/ce5e60672f1b49260f5f8b7060dd029b4cac93f2

**Comment by jezdez** [comment]
@arikfr While we're there, can we add a build argument to install an optional requirements file `requirements_ext.txt` if it exists? That would fix #2810.

File: Dockerfile [hunk 5]
```diff
@@ -66,7 +71,7 @@ ENV PIP_NO_CACHE_DIR=1
 # We first copy only the requirements file, to avoid rebuilding on every file
 # change.
 COPY requirements.txt requirements_bundles.txt requirements_dev.txt requirements_all_ds.txt ./
-RUN pip install -r requirements.txt -r requirements_dev.txt
+RUN if [ "x$skip_dev_deps" = "x" ] ; then pip install -r requirements.txt; else pip install -r requirements.txt -r requirements_dev.txt; fi
```
Comment (arikfr) [inline Dockerfile:74]:
```suggestion
RUN if [ "x$skip_dev_deps" = "x" ] ; then pip install -r requirements.txt -r requirements_dev.txt; else pip install -r requirements.txt; fi
```

**Comment by arikfr** [comment]
> @arikfr While we're there, can we add a build argument to install an optional requirements file `requirements_ext.txt` if it exists? That would fix #2810.

No need for an arg, checking if file exists is enough. But I'm not sure about the usefulness of it. It saves the person to write a single command (`RUN pip install...`) and requires them to understand how the Dockerfile is built... Is it really worth the hassle?

**Comment by jezdez** [comment]
> > @arikfr While we're there, can we add a build argument to install an optional requirements file `requirements_ext.txt` if it exists? That would fix #2810.
> 
> No need for an arg, checking if file exists is enough. But I'm not sure about the usefulness of it. It saves the person to write a single command (`RUN pip install...`) and requires them to understand how the Dockerfile is built... Is it really worth the hassle?

It would be more than just `RUN pip install..` since when inheriting from the Redash image you would have to also switch to the root user before to make sure the extensions are installed in the same location as the rest of the requirements. And then switch back to the `redash` user, which isn't ideal.

```Dockerfile
FROM redash/redash:preview

COPY requirements_ext.txt .
USER root
RUN pip install -r requirements_ext.txt
USER redash
```

But I know realize that there is the devil in the details, since I think ONBUILD is need to make this work in a separate Dockerfile to make sure it happens at the correct time. And ONBUILD isn't enough for frontend extensions to be installed before the webpack bundle is built. Hmm..

Maybe we should stop using webpack in favor of build-less packaging for the frontend assets like [Snowpack](https://snowpack.dev)?

**Comment by arikfr** [comment]
> Maybe we should stop using webpack in favor of build-less packaging for the frontend assets like [Snowpack](https://snowpack.dev)?

Snowpack doesn't require build during development, but you still need to build for production/deployment.

**Pull request merged**